### PR TITLE
Exo keep tries

### PR DESCRIPTION
**Pull request by eldoniel**
This PR fixes the following issue:
https://github.com/claroline/Distribution/issues/531

In formative exercises, the steps gets locked:
- when all tries are used
- when all answers are found

If the user changes step, and then goes back to an already played one, the step will stay locked.

### CODE REVIEW

**Comment by ptsavdar** [comment]
```
console.log('Message: ' + msg)
console.log('Code: ' + code)
console.log('Status: ' + status)
```
I think is better to use the Angular [$log](https://docs.angularjs.org/api/ng/service/$log) service instead of the console.log (I think I saw it in the best practices or something that @stefk sent us the other day)

**Comment by eldoniel** [comment]
There is still one syntax error I can't get rid of. I don't know how to import 'Routing' from fosjsbundle. Any ideas?

Thanks in advance.

**Comment by ptsavdar** [comment]
Have you tried using window.Routing instead of just Routing? I'm not sure it will work but you can try it, if you haven't already tried that.

**Comment by stefk** [comment]
@eldoniel, I don't think it's possible to make a clean import of that lib. It's available globally on every page anyway and we can't change that now, so the only problem is to tell eslint about that global.

I'm pretty sure I mentionned it in the chat, but the best solution IMO is to use a service defined in the core bundle that wraps the router (`main/core/fos-js-router`), so that the global [specification](http://eslint.org/docs/user-guide/configuring#specifying-globals) is done there and you don't have to refer to `Routing` directly.

**Comment by eldoniel** [comment]
@stefk @ptsavdar Thank you for your answers.

I'm not sure I understood well what I had to do. I tried something anyway; it seems to be working and to pass tests better.

File: main/core/Resources/modules/fos-js-router/module.js [hunk 1]
```diff
@@ -1,8 +1,7 @@
...
 import 'angular/angular.min'
 
+import RouterService from './Service/RouterService'
+
 angular.module('ui.fos-js-router', [])
-    .filter('path', function () {
-        return function (route, parameters = {}) {
-            return Routing.generate(route, parameters)
-        };
-    })
```
Comment (stefk) [inline main/core/Resources/modules/fos-js-router/module.js]: Are you sure no existing code use this filter?

File: main/core/Resources/modules/fos-js-router/Service/RouterService.js [hunk 2]
```diff
@@ -0,0 +1,189 @@
+/* global goog */
```
Comment (stefk) [inline main/core/Resources/modules/fos-js-router/Service/RouterService.js]: Where does this code come from ?

File: plugin/exo/Resources/modules/paper/Services/UserPaperService.js [hunk 3]
```diff
@@ -14,6 +14,8 @@ function UserPaperService($http, $q, PaperService, ExerciseService) {
   this.ExerciseService = ExerciseService
 }
 
+/* global Routing */
```
Comment (stefk) [inline plugin/exo/Resources/modules/paper/Services/UserPaperService.js]: The point is to have that annotation in the router service, not here.

File: main/core/Resources/modules/fos-js-router/Service/RouterService.js (hunk 2 above)
Comment (eldoniel) [inline main/core/Resources/modules/fos-js-router/Service/RouterService.js]: I did add it myself, as I was getting a "no-undef" error for this "goog" as well.
(sorry if I do weird things, all of this is really blurry for me right now, I'm a bit lost...)

File: plugin/exo/Manager/ExerciseManager.php [hunk 4]
```diff
@@ -81,7 +81,7 @@ public function deleteStep(Exercise $exercise, Step $step)
 
         // Update steps order
         $steps = $exercise->getSteps();
-        foreach ($steps as $pos => $stepToReorder) {
+        foreach ($steps as $pos) {
```
Comment (Elorfin) [inline plugin/exo/Manager/ExerciseManager.php]: In fact, there is a bug in this method : you need to keep the `$stepToReorder` variables and do `$stepToReorder->setOrder($pos)` (next line).

This method delete the step `$step` and reorder the others. So it's not the deleted one we need to reorder.

File: main/core/Resources/modules/fos-js-router/module.js (hunk 1 above)
Comment (ptsavdar) [inline main/core/Resources/modules/fos-js-router/module.js]: I know I'm using it in website I think and also in portal.

File: plugin/exo/Manager/ExerciseManager.php [hunk 5]
```diff
@@ -271,6 +271,10 @@ public function importExercise($data)
      */
     public function exportExercise(Exercise $exercise, $withSolutions = true)
     {
+        if ($exercise->getType() === '3') {
+            $withSolutions = true;
+        }
```
Comment (Elorfin) [inline plugin/exo/Manager/ExerciseManager.php]: There are constants in `Exercise` entity for type. It's better to use it instead of the direct value.

**Comment by Elorfin** [comment]
LGTM

**Comment by ptsavdar** [comment]
LGTM

**Pull request merged**